### DISCUSSION TITLE
Getting the hash of the last commit when merging

### DISCUSSION THREAD

**Discussion by mdneuzerling** [community/community#25989] [Actions]
<p>I’m using the SHA hash of the last commit as a common value for a few workflows. So when a pull request is made, my PR workflow evaluates <code>{{ github.event.after }}</code> to get this hash, which is not the same as <code>GITHUB_SHA</code>. What I’m retrieving here is the last hash I would see on the GitHub UI before I even make the pull request. This is my common identifier that sets S3 keys, random seeds, etc. So far so good!</p>
<p>When merging the pull request, I have another workflow that triggers on [push]. I’d like to be able to access this same hash. The problem here is that the last commit is now the merge, so really I’m after the second-last commit. My usual tricks with <code>git log</code> don’t seem to work here because the only commit in the log is the merge? When I dump the <code>github</code> context, the has I want exists only in the <code>github.event.commits</code> array, but I’m not sure how to choose the second-last one. Indexing with <code>[-2]</code> doesn’t work, and the array doesn’t have a <code>length</code> attribute).</p>
<p>Is there a way to grab this second-last commit hash on merge? Or, equivalently, a way to grab the last commit hash before the merge? Does this hash have a special name that I’m missing?</p>
<p>I’d be grateful for any help here. I feel like I’m running up against the boundary of my git knowledge, and my practically non-existent Javascript knowledge.</p>

**Comment by weide-zhou** [comment, marked as answer]
<p>Hi <a class="mention" href="/mdneuzerling">@mdneuzerling</a>,</p>
<p>Glad to see you in Github Community Forum!</p>
<p>Since the hash exists in the github.event.commits array, you can use below sample code to get the second-last id, not ideal but it works.</p>
<pre><code class="lang-auto">      - name: get the sha
        run: |
          echo "$GITHUB_CONTEXT"| jq '.event.commits[].id' | tail -2 | head -1 |sed 's/\"//g'
        env:
          GITHUB_CONTEXT: ${{ toJson(github) }}
</code></pre>
<p>Please refer to my workflow for your reference.<br>
</p><aside class="onebox whitelistedgeneric">
  <header class="source">
      <img src="https://github.githubassets.com/favicons/favicon.svg" class="site-icon" width="32" height="32">
      <a href="https://github.com/weide-zhou/ticket13/runs/821121612?check_suite_focus=true#step:3:222" target="_blank" rel="nofollow noopener">GitHub</a>
  </header>
  <article class="onebox-body">
    <img src="https://user-images.githubusercontent.com/52772761/181094446-975428ee-5f7a-4005-b533-898de833f525.png" class="thumbnail onebox-avatar" width="420" height="420">

<h3><a href="https://github.com/weide-zhou/ticket13/runs/821121612?check_suite_focus=true#step:3:222" target="_blank" rel="nofollow noopener">weide-zhou/ticket13</a></h3>

<p>Contribute to weide-zhou/ticket13 development by creating an account on GitHub.</p>


  </article>
  <div class="onebox-metadata">
    
    
  </div>
  <div style="clear: both"></div>
</aside>

<p>Thanks.</p>

**Comment by mdneuzerling** [comment]
<p>Thank you so much, <a class="mention" href="/weide-zhou">@weide-zhou</a>! That use of <code>jq</code> is something I doubt I would have gotten to myself.</p>
<p>For anyone else who comes across this thread in the future:</p>
<ul>
<li>The hash I’m referring to in a pull request is actually <code>github.event.pull_request.head.sha</code>, <strong>not</strong> <code>github.event.after</code>. Under certain circumstances, such as when an open pull request is updated with another commit, these two may be equal.</li>
<li>Another option here is to trigger the workflow when the pull request is closed, rather than on push, and then putting <code>if: github.event.pull_request.merged == true</code> in <em>every</em> job in the workflow. It’s messy, but it’s similar to a “when a pull request is merged” trigger. Then the <code>github.event.pull_request</code> context is available for the workflow.</li>
</ul>

**Comment by ricardofiorani** [comment]
<del>As of 2023, this doesn't work anymore as `github.event` doesn't contain the node `pull_request` or `commits` anymore. 😠<del>

My bad, the `event` will contain a different shape based on how the pipeline was triggered
